src/compiler/semantic/analyser.rs

```diff
@@ -0,0 +1,331 @@
+                    ty: id.ty.clone().map(|ty| ty.name()),
+                };
+
+                RULES.apply_new_symbol_statement(self, param, &id.id, &sym);
+
+                self.symbols
+                    .add_output_variable(&self.cur_scope, id.id.name(), sym);
+            }),
+            _ => unreachable!("parser should only produce signals and vars"),
+        });
+    }
+
+    /// Add state declarations of a machine.
+    /// This is done before analysing the rest of the machine because you can refer to a state in a
+    /// transition before it is declared. This does not happen with other symbols that need to be
+    /// declared before using.
+    fn add_state_decls(&mut self, block: &Statement<BigInt, Identifier>) {
+        if let Statement::Block(_, stmts) = block {
+            stmts.iter().for_each(|stmt| {
+                if let Statement::StateDecl(dsym, id, _) = stmt {
+                    let sym = SymTableEntry {
+                        definition_ref: dsym.clone(),
+                        category: SymbolCategory::State,
+                        ty: None,
+                    };
+
+                    RULES.apply_new_symbol_statement(self, stmt, id, &sym);
+
+                    self.symbols.add_symbol(&self.cur_scope, id.name(), sym);
+                }
+            })
+        } else {
+            unreachable!("the parser should produce machine declaration with a block");
+        }
+    }
+
+    fn analyse_state(&mut self, id: Identifier, block: Statement<BigInt, Identifier>) {
+        self.enter_new_scope(id.name());
+
+        self.analyse_statement(block);
+
+        self.exit_scope();
+    }
+
+    fn analyse_statement(&mut self, stmt: Statement<BigInt, Identifier>) {
+        // Add new variable symbols
+        match stmt.clone() {
+            Statement::SignalDecl(dsym, ids) => ids.into_iter().for_each(|id| {
+                let sym = SymTableEntry {
+                    category: SymbolCategory::Signal,
+                    definition_ref: dsym.clone(),
+                    ty: id.ty.map(|ty| ty.name()),
+                };
+
+                RULES.apply_new_symbol_statement(self, &stmt, &id.id, &sym);
+
+                self.symbols.add_symbol(&self.cur_scope, id.id.name(), sym);
+            }),
+            Statement::WGVarDecl(dsym, ids) => ids.into_iter().for_each(|id| {
+                let sym = SymTableEntry {
+                    category: SymbolCategory::WGVar,
+                    definition_ref: dsym.clone(),
+                    ty: id.ty.map(|ty| ty.name()),
+                };
+
+                RULES.apply_new_symbol_statement(self, &stmt, &id.id, &sym);
+
+                self.symbols.add_symbol(&self.cur_scope, id.id.name(), sym);
+            }),
+            // State decl symbols are added in
+            // add_state_decls
+            Statement::StateDecl(_, _, _) => {}
+            _ => {}
+        }
+
+        RULES.apply_statement(self, &stmt);
+
+        // Recursive
```

Comment:
should we, in the future, worry about this being recursive?

Reply:
Recursive is good, it is just to clarify the different steps that this method makes, perhaps I should split in two new methods